### ISSUE TITLE
search: OR("||") operator support in filter_by clause

### ISSUE THREAD

**Issue by Khatunzev-Anton** [typesense/typesense#281]
## Description

We would like to have OR("||") operator supported inside filter_by statement. At the moment documentation specifies only "&&" operator. Furthermore it would be better to have more complex logical expressions supported(conjunctions of disjunctions for example)

## Steps to reproduce
queries like this `/collections/companies/documents/search?q=*&filter_by=rootfield__document_uri:="di-r_3" || additionaldata_company_competences:="lawyer"` return no results when `/collections/companies/documents/search?q=*&filter_by=rootfield__document_uri:="di-r_3" && additionaldata_company_competences:="lawyer"` has some hits

## Expected Behavior

"||" operator supported in filter_by clause

## Metadata

**Typsense Version**: 2.0

**Comment by kylewill** [comment]
Natural query language (e.g. (michael AND jackson) OR "king of pop") would be core for our use case where false negatives can be high

**Comment by bfritscher** [comment]
Please implement OR. It would be useful for my use case In my algolia setup, I created one searchkey for each user with the following filter `users:${uid} OR public:true`
(users is a string[], we want to allow the user to search for projects in which he participates, or projects which are set as public)

**Comment by alejoar** [comment]
Hey! Came here to share my use case for this. I am currently having the same issue in Algolia.

Let's say I have a DB of "providers", some are local (specific zip code converted to lat/lng e.g. location=lat/lng), some are state wide (they have a set of states as a service area, e.g. "NY", "PA") and some are national (they have an attribute e.g. national=true ).

I would like to be able to let users input a zip code, which my backend would convert to lat/lng coordinates, and then display the user results with a filter like this:

"close to" lat/lng OR state=NY OR national=true

**Comment by jasonbosco** [comment]
Wanted to document a work-around here while we work on #396 which will address this comprehensively. 

Typesense currently only supports `OR`ing values in the same field, and `AND`ing values across fields or the same field. If you need to `OR` across fields, the work-around would be to send multiple queries using the [`multi_search`](https://typesense.org/docs/latest/api/documents.html#federated-multi-search) endpoint, with each query having a particular field in the filter_by expression and then aggregating the results client-side for these queries.

There's a `_text_match` score as part of each `hits` hash that can be used for this cross-collection aggregation.

**Comment by adilansari** [comment]
Is this available already in any 0.24 rc?

**Comment by jasonbosco** [comment]
Not yet, but it's being actively worked on.

**Comment by kishorenc** [comment]
We now have a Docker build which supports cross field ORs:

```
typesense/typesense:0.24.0.rcn17
```

We've added a `||` operator for doing ORs. So, you can now do something like: `grade:>10 || is_premier: true`

cc @kiloki-official

**Comment by s-gal** [comment]
@kishorenc 

That's great. Is it possible to put brackets in the statements if you have a more complex scenario like:

(grade:>10 || is_premier: true) && gender:="female"

**Comment by kishorenc** [comment]
Yes, nested filter expressions are supported in this build as well. I will update my comment above to clarify that as well.

**Comment by s-gal** [comment]
Should this feature also work in combination with scoped search keys? I have a problem and do not think the error is on my side...

I defined a scoped key for a user and set the filterBy attribute something like this for the scoped key: filterBy: `(gender: Female) || (grade_math:=1 || grade_english:=2)`... (i know the brackets here are not needed but for some reasons of how my program works i leave them there)

Now if i use the multi_search endpoint with that generated scoped key and the filter_by attribute gets set there the result gets not filtered correctly anymore....

If i set here the filter to grade_math:=1 i would assume all results containing grade_english:=2 would show up only if they would have set also grade_math:=1... But even some results show up that have grade_math not set to 1... Not all though which is strange and i can not figure out what the problem is here and assume its somewhere hidden inside the typesense lib

**Comment by kishorenc** [comment]
> If i set here the filter to grade_math:=1 i would assume all results containing grade_english:=2 would show up only if they would have set also grade_math:=1...

`(grade_math:=1 || grade_english:=2)` filter says: fetch a record that has either `grade_math: 1` or `grade_english: 2`. To achieve what you want, you need to have: `(grade_math:=1 && grade_english:=2)`.

If I'm now following you correctly, please share a small code snippet with sample data + query that show cases the issue.

**Comment by s-gal** [comment]
@kishorenc yes the filter in the scoped search key is something like: `(grade_math:=1 || grade_english:=2)`

But now my problem is, if i set the filter_by attribute inside the multi_search endpoint (and using this scoped search key defined above) just to `grade_math:=1` my assumption would be that the result is further reduced the documents with only grade_math=1.

But if i do so also some records show up that have grade_math !=1 but grade_english =2...

So my guess is something is wrong using the filter_by with OR operators inside scoped keys and trying to further reduce the results with the filter_by attribute of the multi_search endpoint.

```
client.keys().generateScopedSearchKey(keyWithSearchPermissions, {'filter_by': '(grade_math:=1 || grade_english:=2)'}

let searchRequests = {
  'searches': [
    {
      'collection': 'grades',
      'q': '*',
      'filter_by': 'grade_math:=1'
    }
  ]
}

// Search parameters that are common to all searches go here
let commonSearchParams =  {
    'query_by': 'grade_math',
}

client.multiSearch.perform(searchRequests, commonSearchParams)

Result:

id | grade_math | grade_english
1 | 1 | 2
2 | 2 | 2
3 | 1 | 2

-- So result with id = 2 should not show up
```

(So my example is in fact bigger more complex, but should be reproducable like this, and i using a lot of facets, but also like the total result numbers do not much with the facet result numbers displayed)

**Comment by kishorenc** [comment]
@s-gal 

I'm not able to reproduce that issue. Here's a detailed example that I've put together here: https://gist.github.com/kishorenc/5fa4969ba2a27fc040a9dd9ac53fe4d2

See the last line: only the expected document is returned and not both. If you can put together a full example like (in any client of your choice) then it will be easier to debug what's going wrong.

**Comment by kishorenc** [comment]
Alternatively if you are able to share your dataset with us, we can take a look, but I understand it might not be possible. If it is, please email us: contact@typesense.org

**Comment by s-gal** [comment]
@kishorenc thanks for the idea, but yes i am working for a company and its sensitive data and i am not allowed to share

**Comment by kishorenc** [comment]
Since we are dealing with just a couple of integer fields here, you could just rename them as `fieldA` and `fieldB` and then only export those two fields out for the debugging? That will require sharing no real data except a distribution of numbers which will be no of no use to anyone.

**Comment by s-gal** [comment]
@kishorenc 
the collection i have, has over 100 data fields in all kind of different types and the collection is over 50000 entries big... Not sure exacly if i leave data away you will face the same issues as i do... as said maybe later this week i have some more time and can test/prepare something to be looked at

**Comment by s-gal** [comment]
@kishorenc 
i just reproduced the problem and sent you all needed data and informations to reproduce my problem to the email address you mentioned....

**Comment by kishorenc** [comment]
Thanks, I'll take a look and update.

**Comment by kishorenc** [comment]
Fixed in `v0.24`
